src/Ast/Type/ArrayShapeNode.php

```diff
@@ -13,7 +14,7 @@ class ArrayShapeNode implements TypeNode
 
 	use NodeAttributes;
 
-	/** @var ArrayShapeItemNode[] */
+	/** @var array<ArrayShapeItemNode|CommentNode|IdentifierTypeNode> */
```

Comment:
I don't think comments should be part of the AST as nodes. Being in node attributes would match how `nikic/PHP-Parser` works, and I would like that here too. It would make it easier to work with it, and also support comments everywhere, without BC breaks.

Reply:
Okay, but can you please provide more of a big picture to me, here? ie:
1. why do we need to retain them at all?
2. does the printer need to reproduce them?
3. if so, what about new lines? Single line comments can depend on new lines to prevent them from affecting following content

And a little background about the printer would be helpful; if that's producing the output of `dumpType` and what we see in error messages, then I imagine you would still want single-line output, in which case comments are just going to add noise, and if we agree on that then it's not clear to me why we shouldn't just ignore comments.

Reply:
I took another crack at it:

* replaced simple // regex with a more robust pattern that captures single and block comments (so they can happen anywhere, as you say)
* modified `parseArrayShape` to catch comments. This seems to be necessary because array shape is one of the few node types that does recursive parsing (I had a look at nikic parser, by the way, and they have an architecture that works better for this kind of thing, with all the work happening in a centralized `next` method that drives the entire process).

Am I getting warmer?

Reply:
I had a look through NIkic's printer; the section dealing with comments spans about 100 lines. As you say, it's vaguely similar to your code. I figure I'm looking at about a week or more of debugging and trying to understand what it means to add/remove shape items, and work out all the kinks. If that's going to be a blocker on this PR then I guess I'll have no choice, but given that the tests pass as-is, is there any way we can leave that stuff for another time?

Reply:
@ondrejmirtes Sorry to be a pest, but I'm blocked by this PR in my other work. Do you require the format-preserving printer stuff to move forward with this PR?

Reply:
Yes, it's a blocker, otherwise it'd mean I have to implement it myself after merging the PR without it.

This is a serious low-level library, I'd expect an issue here or in Slevomat CS pretty soon after people would start commenting their PHPDoc types.

Reply:
Very well. I'll dive in on it, thanks much.

Reply:
Made some progress. Still not sure I understand what I'm doing, but I've added a new printer test that uses the visitor pattern to add a few `ArrayShapeItems` with comments to a multiline array shape, and got the `DiffElem::TYPE_ADD` branch of `Printer::printArrayFormatPreserving` working, at least with single line comments. Whew.

Tomorrow I'll see what I can do about multi-line comments, but in the meantime can you shed any light on these CI errors?

```
[BC] ADDED: Method getComments() was added to interface PHPStan\PhpDocParser\Ast\Node
[BC] ADDED: Method getComments() was added to interface PHPStan\PhpDocParser\Ast\Type\TypeNode
[BC] ADDED: Method getComments() was added to interface PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTagValueNode
[BC] ADDED: Method getComments() was added to interface PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocChildNode
[BC] ADDED: Method getComments() was added to interface PHPStan\PhpDocParser\Ast\ConstExpr\ConstExprNode
```

Thanks much.

Reply:
You can't add methods on interfaces, that's a BC break in case someone implements their own Node.

You always need to access the comments through getAttribute.

Reply:
I see, thank you. Updated.